File: java/25.java
// 25. Reverse Nodes in k-Group
// Given a linked list, reverse the nodes of a linked list k at a time and return its modified list.
//
// k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes, in the end, should remain as it is.
//
// You may not alter the values in the list's nodes, only nodes themselves may be changed.
//
// Example 1:
//
//
// Input: head = [1,2,3,4,5], k = 2
// Output: [2,1,4,3,5]
// Example 2:
//
//
// Input: head = [1,2,3,4,5], k = 3
// Output: [3,2,1,4,5]
// Example 3:
//
// Input: head = [1,2,3,4,5], k = 1
// Output: [1,2,3,4,5]
// Example 4:
//
// Input: head = [1], k = 1
// Output: [1]
//
// Constraints:
//
// The number of nodes in the list is in the range sz.
// 1 <= sz <= 5000
// 0 <= Node.val <= 1000
// 1 <= k <= sz
//
// Runtime: 1 ms, faster than 20.41% of Java online submissions for Reverse Nodes in k-Group.
// Memory Usage: 42.6 MB, less than 5.31% of Java online submissions for Reverse Nodes in k-Group.
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k == 1) return head;
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode cur = head;
        int count = 1;
        ListNode prevTail = dummy;
        while(cur != null) {
            if (count == k) {
                ListNode nextHead = cur.next;
                cur = head;
                ListNode node  = head.next;
                head.next = nextHead;
                while(node != nextHead) {
                    ListNode temp = node.next;
                    node.next = head;
                    head = node;
                    node = temp;
                }
                prevTail.next = head;
                prevTail = cur;
                head = nextHead;
                cur = nextHead;
                count = 1;
            } else {
                cur = cur.next;
                count++;
            }
        }
        return dummy.next;
    }
}

File: go/25.go
// 25. Reverse Nodes in k-Group

// Given the head of a linked list, reverse the nodes of the list k at a time, and return the modified list.

// k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes, in the end, should remain as it is.

// You may not alter the values in the list's nodes, only nodes themselves may be changed.

 

// Example 1:


// Input: head = [1,2,3,4,5], k = 2
// Output: [2,1,4,3,5]
// Example 2:


// Input: head = [1,2,3,4,5], k = 3
// Output: [3,2,1,4,5]
 

// Constraints:

// The number of nodes in the list is n.
// 1 <= k <= n <= 5000
// 0 <= Node.val <= 1000
 

// Follow-up: Can you solve the problem in O(1) extra memory space?

// Runtime 0 ms Beats 100.00%
// Memory 5.36 MB Beats 86.91%
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
    if k == 1 {
        return head
    }
    dummyHead := &ListNode{Next: head}
    tempHead := dummyHead
    first := tempHead.Next
    for ; first != nil; {
        p1 := first
        for i := 0; i < k; i++ {
            if p1 == nil {
                return dummyHead.Next
            }
            p1 = p1.Next
        }
        cur := first
        tempNext := cur
        for i := 0; i < k; i++ {
            if cur == nil {
                first.Next = nil
                return dummyHead.Next
            }
            tempNext = cur.Next
            cur.Next = tempHead.Next
            tempHead.Next = cur
            if (i == k - 1) {
                tempHead = cur
            }
            cur = tempNext
        }
        first.Next = p1
        tempHead = first
        first = first.Next
    }
    return dummyHead.Next
}